rocketpy/tools.py

```diff
@@ -23,3 +26,978 @@ def __get__(self, instance, owner=None):
+
+        Examples
+        --------
+        >>> M = Matrix([[lambda x: x**1, lambda x: x**2, lambda x: x**3],
+        ...             [lambda x: x**4, lambda x: x**5, lambda x: x**6],
+        ...             [lambda x: x**7, lambda x: x**8, lambda x: x**9]])
+        >>> M(2)
+        Matrix([2, 4, 8],
+               [16, 32, 64],
+               [128, 256, 512])
+        """
+        try:
+            return self.element_wise(lambda f: f(*args))
+        except TypeError as exc:
+            msg = "One or more elements of this matrix is not callable."
+            raise TypeError(msg) from exc
+
+    def __len__(self):
+        """Adds support for the len() function."""
+        return 3
+
+    @cached_property
+    def shape(self):
+        """tuple: Shape of the matrix."""
+        return (3, 3)
+
+    @cached_property
+    def xx(self):
+        return self[0, 0]
+
+    @cached_property
+    def xy(self):
+        return self[0, 1]
+
+    @cached_property
+    def xz(self):
+        return self[0, 2]
+
+    @cached_property
+    def yx(self):
+        return self[1, 0]
+
+    @cached_property
+    def yy(self):
+        return self[1, 1]
+
+    @cached_property
+    def yz(self):
+        return self[1, 2]
+
+    @cached_property
+    def zx(self):
+        return self[2, 0]
+
+    @cached_property
+    def zy(self):
+        return self[2, 1]
+
+    @cached_property
+    def zz(self):
+        return self[2, 2]
+
+    @cached_property
+    def trace(self):
+        """Matrix trace, sum of its diagonal components."""
+        return self.xx + self.yy + self.zz
+
+    @cached_property
+    def transpose(self):
+        """Matrix transpose."""
+        return Matrix(
+            [
+                [self.xx, self.yx, self.zx],
+                [self.xy, self.yy, self.zy],
+                [self.xz, self.yz, self.zz],
+            ]
+        )
+
+    @cached_property
+    def det(self):
+        """Matrix determinant."""
+        return self.__abs__()
+
+    @cached_property
+    def is_diagonal(self, tol=1e-6):
+        """Boolean indicating if matrix is diagonal.
+
+        Parameters
+        ----------
+        tol : float, optional
+            Tolerance used to determine if non-diagonal elements are negligible.
+            Defaults to 1e-6.
+
+        Returns
+        -------
+        bool
+            True if matrix is diagonal, False otherwise.
+
+        Examples
+        --------
+        >>> M = Matrix([[1, 0, 0], [0, 2, 0], [0, 0, 3]])
+        >>> M.is_diagonal
+        True
+
+        >>> M = Matrix([[1, 0, 0], [0, 2, 0], [0, 1e-7, 3]])
+        >>> M.is_diagonal
+        True
+
+        >>> M = Matrix([[1, 0, 0], [0, 2, 0], [0, 1e-5, 3]])
+        >>> M.is_diagonal
+        False
+        """
+        for i, j in product(range(3), range(3)):
+            if i == j:
+                continue
+            if abs(self[i, j]) > tol:
+                return False
+        return True
```

Comment:
Just as a remider (to myself or anyone who wants to toy with timeit) to compare the speed of this implementation and an huge explicit `if` with `ands` (since Python evaluates them lazily). I really hope its not much.

Reply:
Just checked and the results are remarkebly close and within run to run variance. Therefore this implementation is much better.